src/app/concretes/QuorumAppImpl.sol

```diff
@@ -0,0 +1,154 @@
+                _validatorIdByAddress[validator] = id;
+                _validatorAddressById[id] = validator;
+            }
+        }
+        require(validatorCount >= 1, "quorum must not be empty");
+        _VALIDATOR_COUNT = validatorCount;
+    }
+
+    function getGenesisStateRoot()
+        public
+        view
+        override
+        returns (bytes32 genesisStateRoot)
+    {
+        return _GENESIS_STATE_ROOT;
+    }
+
+    function getEpochFinalizerInterfaceId() external pure override returns (bytes4) {
+        return type(Quorum).interfaceId;
+    }
+
+    function closeEpoch(uint256 epochIndex) external override {
+        canEpochBeClosed(epochIndex);
+        _closeEpoch(address(this));
+    }
+
+    function finalizeEpoch(
+        uint256 epochIndex,
+        bytes32 postEpochOutputsRoot,
+        bytes32[] calldata proof
+    ) external override {
+        bytes32 postEpochStateRoot;
+        postEpochStateRoot = _preFinalize(epochIndex, postEpochOutputsRoot, proof);
+        _finalizeEpoch(postEpochStateRoot, postEpochOutputsRoot);
+    }
+
+    function getValidatorCount() public view override returns (uint8 validatorCount) {
+        validatorCount = _VALIDATOR_COUNT;
+    }
+
+    function getValidatorAddressById(uint256 validatorId)
+        external
+        view
+        override
+        returns (address validatorAddress)
+    {
+        validatorAddress = _validatorAddressById[validatorId];
+    }
+
+    function getValidatorIdByAddress(address validatorAddress)
+        public
+        view
+        override
+        returns (uint8 validatorId)
+    {
+        validatorId = _validatorIdByAddress[validatorAddress];
+    }
+
+    function getVoteBitmap(uint256 epochIndex, bytes32 postEpochStateRoot)
+        public
+        view
+        override
+        returns (bytes32 voteBitmap)
+    {
+        voteBitmap = _voteBitmaps[epochIndex][postEpochStateRoot];
+    }
+
+    function getAggregatedVoteBitmap(uint256 epochIndex)
+        public
+        view
+        override
+        returns (bytes32 aggregatedVoteBitmap)
+    {
+        aggregatedVoteBitmap = _aggregatedVoteBitmaps[epochIndex];
+    }
+
+    function vote(uint256 epochIndex, bytes32 postEpochStateRoot)
+        external
+        override
+        isFirstNonFinalizedEpoch(epochIndex)
+    {
+        require(_isFirstNonFinalizedEpochClosed(), CannotCastVoteForOpenEpoch());
+
+        // First, make sure the message sender is a validator.
+        address validatorAddress = msg.sender;
+        uint8 validatorId = getValidatorIdByAddress(validatorAddress);
+        require(validatorId != 0, MessageSenderIsNotValidator(validatorAddress));
+
+        // Second, check whether the validator has already voted in the epoch.
+        // If not, mark the validator as having already voted in the epoch.
+        {
+            bytes32 bitmap = getAggregatedVoteBitmap(epochIndex);
+            require(!bitmap.getBitAt(validatorId), VoteAlreadyCastForEpoch());
```

Comment:
is this error necessary?

Reply:
This check ensures `Vote` is only emitted on new votes.
It may be misleading to have a `Vote` event for a non-change.

Reply:
Yes, actually this error ensures that validators cannot vote on multiple post-epoch state roots for a given epoch.